core: make targetManager accessible on driver

brendankenny opened the pull request:
Should help significantly with #14078. Smoke tests should all pass, but waiting on feedback before touching the unit test mocks :)

This PR elevates `TargetManager` to a top-level property on `Driver`. The intention is that TargetManager
- will be the sole place tracking and attaching to targets/sessions (including wrapping `CDPSession`s in FR `session`s), which should make the execution flow easier to understand and make future changes simpler.
- will live for as long as `Driver` does, which includes across passes. It's basically a live view of all the sessions Lighthouse knows about
- the place to subscribe to all protocol events (with the flattened `sessionId` included)

The legacy driver basically does these things already, so it was easy to expose a shim on it and keep `NetworkMonitor` (and its users) on a unified implementation. It's not pretty architecturally, but it is simple (yay) and should only be around until LH 11.

Since `TargetManager` tracks all sessions and their targetInfo, in the future it could be used for things like, "give me all the iframe sessions" so you can send commands or `evaluate()` in them easily, but nothing needs that right now and it's harder to include legacy support, so we can wait.

re: @paulirish's https://github.com/GoogleChrome/lighthouse/pull/14120#issuecomment-1154359719, AFAICT _all_ sessions now have IDs, presumably because pptr is keeping the `sessionId`-less session to itself. I'd like to verify that and move us to a place where `sessionId` is required in the all-protocol-events type:

https://github.com/GoogleChrome/lighthouse/blob/1d6ee0ef0fde4acbdae0dbbe67de7d1e5d0ac5fa/types/protocol.d.ts#L68-L69

so code will hopefully be forced to contend with it. Meanwhile in this PR, `session` drops the `sessionId` so it's like `CDPSession`. If you're subscribing to a single session's events, there's no need for a session ID. We can follow up and make the `session` event handling simpler, but I didn't want the changes in this PR interleaved with that change.

brendankenny commented on lighthouse-core/fraggle-rock/gather/driver.js:
```diff
@@ -67,8 +72,10 @@ class Driver {
     log.time(status);
-    const session = await this._page.target().createCDPSession();
-    this.defaultSession = new ProtocolSession(session);
+    const cdpSession = await this._page.target().createCDPSession();
+    this._targetManager = new TargetManager(cdpSession);
+    await this._targetManager.enable();
+    this.defaultSession = this._targetManager.rootSession();
```
Because `TargetManager` is in charge of creating FR `ProtocolSession`s from `CDPSession`s, and we want to eventually make those sessions available for gatherers to access directly, Driver now gets its `driver.defaultSession` from targetManager instead of making it itself. This is a little weird, but otherwise we have two different session objects pointing to the same thing. This seemed conceptually simpler (the set of all active sessions includes `defaultSession`) but open to ideas.

paulirish replied:
Big fan of avoiding the double session object! (I was seeing that at the CDP traffic level. so weird.)

brendankenny commented on types/protocol.d.ts:
```diff
@@ -5,6 +5,23 @@
+   *   'Domain.method2Name': {method: 'Domain.method2Name', params: EventPayload2},
+   * }
+   */
+  type RawEventMessageRecord = {
```
this is an unchanged move from below, but it was previously scoped to the file and we need it exported now

brendankenny commented on lighthouse-core/gather/gatherers/devtools-log.js:
```diff
@@ -42,16 +40,16 @@ class DevtoolsLog extends FRGatherer {
-    this._networkMonitor.on('protocolmessage', this._onProtocolMessage);
-    this._networkMonitor.enable();
+    driver.targetManager.on('protocolevent', this._onProtocolMessage);
+    await driver.defaultSession.sendCommand('Page.enable');
```
we have so many `Page.enable`s in LH that this has to be unnecessary, but this matches the behavior of the `this._networkMonitor.enable()` call

connorjclark commented on lighthouse-core/gather/driver/target-manager.js:
```diff
@@ -5,70 +5,125 @@
+     * A map of target id to target/session information. Used to ensure unique
+     * attached targets.
+     * @type {Map<string, TargetWithSession>}
+     * */
```
```suggestion
     */
```

connorjclark commented on lighthouse-core/gather/driver.js:
```diff
@@ -96,6 +96,29 @@ class Driver {
     /** @private @deprecated Only available for plugin backcompat. */
     this.evaluateAsync = this.executionContext.evaluateAsync.bind(this.executionContext);
+
+    // A shim for sufficient coverage of targetManager functionality.
```
more comment, maybe:

`// Legacy driver does not need target management because the legacy driver explicitly handles Target.attachedToTarget events. See _handleTargetAttached`

paulirish commented on lighthouse-core/fraggle-rock/gather/driver.js:
```diff
@@ -67,8 +72,10 @@ class Driver {
+    const cdpSession = await this._page.target().createCDPSession();
+    this._targetManager = new TargetManager(cdpSession);
+    await this._targetManager.enable();
+    this.defaultSession = this._targetManager.rootSession();
```
i'd love to s/defaultSession/rootSession/g but fine with it happening in a followup.

connorjclark commented on lighthouse-core/gather/driver/target-manager.js:
```diff
@@ -5,70 +5,125 @@
 
-class TargetManager {
+/**
+ * Tracks targets (the page itself, its iframes, their iframes, etc) as they
```
\+ workers? I know etc covers that...but still

brendankenny replied:
technically we [don't attach to those (yet)](https://github.com/GoogleChrome/lighthouse/pull/14122/files#diff-1d958bf77bb0afbc572b7eb91124d9adeec7802edfbf4fe59837ebf2e48a7e26R96) :)

connorjclark commented on lighthouse-core/gather/driver/target-manager.js:
```diff
@@ -5,70 +5,125 @@
-        waitForDebuggerOnStart: true,
-      });
-    };
+    throw new Error(`session not found`);
```
```suggestion
    throw new Error(`session ${sessionId} not found`);
```

connorjclark reviewed (commented):
(first pass)

paulirish commented on lighthouse-core/gather/driver/target-manager.js:
```diff
@@ -5,70 +5,125 @@
 
-      await session.sendCommand('Target.setAutoAttach', {
+      // We want to receive information about network requests from iframes, so enable the Network domain.
+      await newSession.sendCommand('Network.enable');
```
do you think we should also tweak this to do a promise.all style? https://github.com/GoogleChrome/lighthouse/commit/4bc0b052c187aea0f23ffd704606fa5c0581a7c5
(im still surprised this is fine without it)
also.. i wonder if the `finally` case hits because of a protocol_timeout on one of these?

i know you saw some patterns in pptr/playwright which felt more confident around this stuff. dunno if that's relevant.

paulirish reviewed (commented):
lots of great improvements here. thanks for taking this on.

adamraine commented on lighthouse-core/gather/gatherers/devtools-log.js:
```diff
@@ -11,9 +11,10 @@
-const NetworkMonitor = require('../driver/network-monitor.js');
 const FRGatherer = require('../../fraggle-rock/gather/base-gatherer.js');
 
+/** @typedef {import('../driver/target-manager.js')} TargetManager */
```
Where is this used?

brendankenny replied:
> Where is this used?

it's not, deleted :)

brendankenny commented:
Addressed feedback and unit tests updated.

Let's add a mock simplification pass to our project TODOs, that was way harder than refactoring the actual code :)

connorjclark commented:
This seems to close out the linked issue. Am I missing something?
